Added support for ONNX flavor

avflor opened the pull request:
Addresses #1117

avflor committed:
Added support for ONNX flavor

smurching commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+    print('Requires Python 3.6')
+    sys.exit(1)
+
+import os
```
It's kind of gross, but it seems the best solution to eliminating linter errors here is to include a "  # noqa" or "  #  pylint:disable=E402" after each import line as described in https://github.com/PyCQA/pycodestyle/issues/264

smurching replied:
Note that you'll also want to ensure these tests run in Travis only against Python 3. You can do this by 1) marking the tests as large via the `@pytest.mark.large` annotation before the test declaration (see [here](https://github.com/mlflow/mlflow/blob/7d71ff3894699b5800e52080c8c0f6fe5d46832e/tests/pyfunc/test_spark.py#L68) for an example) and 2) adding a line to the end of .travis.yml (under the "script" section) that looks like:

```
  - if [[ "$TRAVIS_PYTHON_VERSION" == "3.6" ]]; then
      pytest --verbose tests/onnx --large
    fi 
```

dbczumar commented on tests/onnx/test_onnx_model_export.py:
```diff
@@ -0,0 +1,198 @@
+            model_loaded = mlflow.onnx.load_model(
+                "onnx_model",
+                run_id=mlflow.active_run().info.run_uuid)
+            onnx.checker.check_model(model_loaded)
```
The `mlflow.onnx.load_model` method already performs the `check_model` test. We should verify that this is the case by mocking the `check_model` method and asserting that the mock is invoked when calling `mlflow.onnx.load_model`. If another test already verifies that `mlflow.onnx.load_model` calls `check_model`, then we can remove this line.

avflor replied:
Changed tests to use mock.

dbczumar commented on .travis.yml:
```diff
@@ -91,12 +91,15 @@ script:
   - pytest --verbose tests/h2o --large
   # TODO(smurching): Re-enable Keras tests once they're no longer flaky
   # - pytest --verbose tests/keras --large
+  - if [[ "$TRAVIS_PYTHON_VERSION" == "3.6" ]]; then
```
I don't think that this is necessary. The `pytestmark` setting in the test suite should automatically skip the tests when running against Python 2.

avflor replied:
I removed the pytest mark from the tests and kept the if statement in travis scripts.

avflor replied:
Added back the `pytestmark `  and removed the if condition from travis

dbczumar reviewed (changes requested):
@avflor Thank you for this contribution! I've left a few comments regarding Python version restrictions, dependencies, onnx APIs, prediction semantics, and tests. Please let me know if you have any feedback or questions regarding these comments by replying inline.

@smurching Please take a look as well and let me know if you have pushback or questions regarding any of the comments.

avflor commented:
@dbczumar Thank you very much for the suggestions! A few of your comments will be addressed in my latest update. I am still making changes to make sure the code runs with both python 2.7 and 3.6. I will reply asap to the rest of your comments.

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+    model_data_path = os.path.join(path, model_data_subpath)
+
+    # Save onnx-model
+    onnxmltools.utils.save_model(onnx_model, model_data_path)
```
Thanks for the pointer-- onnx.save_model seems to work as expected. I removed the onnxmltools dependency from the code.

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,182 @@
+    def __init__(self, path):
+        self.onnx_model = _load_model(path)
+        self.rt = onnxrt.InferenceSession(path)
+        assert len(self.rt.get_inputs()) == 1
```
@dbczumar It does make sense. However, when I was testing the code with similar setup as the keras tests (iris dataset), I noticed that for this dataset `onnxruntime.get_inputs()` returns only one input (['dense_1_input_01']) which corresponds to the whole dataframe that we pass as input to `predict` in the tests. Thus, for this dataframe, the  `feed_dict = {
            input_name: dataframe[input_name].values
            for input_name in self.input_names
        }` fails. (The dataframe does not include a 'dense_1_input_01' column).
 For this reason, I made it a requirement that there should be only one input as `predict` accepts one dataframe as input. Any thoughts?

dbczumar replied:
@avflor This is a good point. Based on the suggestion you provided in your other comment, I think it makes sense to implement the following behavior for pyfunc inference:

1. In the case where the underlying ONNX model only has a single input, pass `df.values` to the ONNX model (as in the current state of the PR).

2. In the case where the underlying ONNX model has *multiple* inputs, attempt to select dataframe columns with the same names as the ONNX model inputs (as in my proposed solution above).

This proposal has the benefit of supporting multi-input ONNX models while maintaining ease of use for the most common case where the model has a single input.

I also considered the following alternatives:

1. **Keep the current state of the `_OnnxModelWrapper`, restricting the `pyfunc` representation to ONNX models with only one input**. In this case, it does not make sense to add the `pyfunc` flavor to ONNX models with multiple inputs because the `pyfunc` representation cannot be used to perform inference. However, it also seems confusing to omit the `pyfunc` flavor in these cases because MLflow will produce different sets of flavors for a model based on the number of ONNX inputs. This approach also has the disadvantage of restricting interoperability for a broad class of multi-input models.

2. **Use the tensorflow-style, column-based approach for evaluating all ONNX models as python functions, regardless of the number of ONNX inputs**. This API makes it challenging for users to feed multi-dimensional input values to models (e.g., an n-by-4 matrix of Iris data); it requires the user to embed a multi-dimensional array inside a dataframe column. While the current proposal still requires this nested list behavior for ONNX models with *multiple* inputs, the proposed `df.values` API for ONNX models with one input provides a much simpler API for specifying multidimensional data. Additionally, the column-based approach has the disadvantage of always requiring the user to find the names of all ONNX inputs for a model, even if the model only contains a single input. As you noted in the case of converted Keras models, ONNX input names are often autogenerated by various conversion APIs and are not particularly intuitive.

Please let me know if you have any questions or pushback about the proposed solution and the alternatives considered. If we agree to adopt this approach, it's important to thoroughly document the evaluation behavior and provide examples.

avflor replied:
@dbczumar I'm also in favor of the approach you proposed as it allows supporting a broader range of models without making the API more complex. I added both single-input and multi-input implementations during inference. Please take a look and let me know if you have any other comments.

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+
+    def predict(self, dataframe):
+        values = dataframe.values
+        if values.dtype == np.float64:
```
It is interesting that in this case the multiple inputs implementation that you suggested works. I'm thinking maybe this has to do with the implementation of the converters in onnxmltools. Ideally, we should support all converters seamlessly. Should we keep both (single input, multiple input implementations) and specify a condition that picks the correct one? (Maybe examine whether the `onnx_model.input_names` are included in the dataframe or not? It seems ugly but can't think of another way. Any suggestions?

avflor replied:
@dbczumar I forgot to address this one before. The current test (conversion from keras) fails without the conversion but I pushed the conversion to the test and removed it from the inference part.

avflor commented on tests/onnx/test_onnx_model_export.py:
```diff
@@ -0,0 +1,191 @@
+mlflow.onnx = pytest.importorskip('mlflow.onnx')
+
+pytestmark = pytest.mark.skipif(
+    (sys.version_info < (3, 6)),
```
I can do that but since mlflow is tested with Python 3.6, I thought it's better to specify that version.

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+    additional_conda_deps=None,
+    additional_pip_deps=[
+        "onnx=={}".format(onnx.__version__),
+        "onnxmltools=={}".format(onnxmltools.__version__),
```
Removed

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+                            'dependencies': [
+                                'python=3.6.0',
+                                'onnx=1.4.1',
+                                'onnxmltools=1.4.0',
```
Removed

avflor commented on test-requirements.txt:
```diff
@@ -6,6 +6,10 @@ botocore<=1.12.84
+onnx==1.4.1; python_version >= "3.0"
+onnxmltools==1.4.0; python_version >= "3.0"
+onnxruntime==0.3.0; python_version >= "3.0"
+onnxtk==0.0.1; python_version >= "3.0"
```
Removed

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+                            'dependencies': [
+                                'python=3.6.0',
+                                'onnx=1.4.1',
+                                'onnxmltools=1.4.0',
```
Removed

avflor commented on tests/onnx/test_onnx_model_export.py:
```diff
@@ -0,0 +1,198 @@
+    assert conda_env == mlflow.onnx.DEFAULT_CONDA_ENV
+
+
+@pytest.mark.large
```
I added the large marks in the tests based on @smurching suggestions to make it easy to run the tests using python 3.6 only. This seems to work well now. Is there another approach that we should try?

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+        path = mlflow.tracking.artifact_utils._get_model_log_dir(model_name=path, run_id=run_id)
+    path = os.path.abspath(path)
+    flavor_conf = _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME)
+    # Flavor configurations for models saved in MLflow version <= 0.8.0 may not contain a
```
Done. Thanks for the explanation.

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+from __future__ import absolute_import
```
Created issue #1204

avflor commented on tests/onnx/test_onnx_model_export.py:
```diff
@@ -0,0 +1,198 @@
+
+    # Loading ONNX model
+    model_loaded = mlflow.onnx.load_model(model_path)
+    onnx.checker.check_model(model_loaded)
```
Changed tests to use mock.

avflor commented:
@dbczumar @smurching I made a pass over the comments and changed my code accordingly. I have left a few questions/comments in the discussion. Please let me know your input.

avflor commented:
@dbczumar @smurching Any comments? I would really like to move forward with this PR asap. Looking forward to your feedback!

dbczumar reviewed (changes requested):
@avflor Sorry for the delay - I'll try my best to reduce review latency going forward. I've responded to your questions regarding the behavior of the `pyfunc` model representation and highlighted a few outstanding comments that need to be addressed. Please let me know if you have any questions.

avflor commented on travis/run-large-python-tests.sh:
```diff
@@ -18,12 +18,16 @@ trap 'err=1' ERR
 # - pytest --verbose tests/keras --large
+if [[ "$TRAVIS_PYTHON_VERSION" == "3.6" ]]
+then
+   pytest --verbose tests/onnx --large;
```
Added back the pytestmark and removed the if condition from travis

avflor commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,175 @@
+
+import sys
+
+if sys.version_info <= (3, 6):
```
@dbczumar Done

avflor commented:
@dbczumar Thank you so much for the comments. I addressed them --short summary:

1) Kept both single and multiple input implementation in `predict` as you suggested. I also think this is the best way to support multiple ONNX models.
2) Removed python version check from onnx.py and used `pytestmark` in the tests. 
3) Type conversion outside the `predict ` function does not seem to work due to an issue with JSON serialization in the scoring server. Please see comments above and issue #1286.

Please let me know if you have any more comments/suggestions.

dbczumar commented:
@avflor Thank you for your updates! I'll take another pass within 24 hours!

dbczumar commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,181 @@
+                            ]
+                        }
+
+    :param kwargs: kwargs to pass to ``onnx_model.save`` method.
```
This `kwargs` parameter does not seem to be defined on `mlflow.onnx.save_model`. Additionally, the [onnx.save_model](https://github.com/onnx/onnx/blob/51b6ecce7a6f7bfc3b7944e77614bc0c38d0db96/onnx/__init__.py#L168) method doesn't take any keyword arguments that are currently used; here is the signature:
```
def save_model(proto, f, format=None):  # type: (Union[ModelProto, bytes], Union[IO[bytes], Text], Optional[Any]) -> None
    '''
    Saves the ModelProto to the specified path.
    @params
    proto should be a in-memory ModelProto
    f can be a file-like object (has "write" function) or a string containing a file name
    format is for future use
    '''
```

I think it makes sense to remove the `kwargs` parameter for now.

dbczumar reviewed (changes requested):
@avflor Thanks for addressing the previous round of comments! This is very close. I left a few additional comments about testing multi-input model scoring with `pyfunc`, removal of casting behavior, and documentation.

avflor commented:
@dbczumar I have addressed all the comments (added test for multi-input models, incorporated the workaround we discussed in `predict`
and added imports directly in the tests). Please let me know if you have any comments. 

I have some trouble with the R tests though. I have rebased over master but the R tests keep failing with the following error:
```
══ testthat results  ═══════════════════════════════════════════════════════════

OK: 189 SKIPPED: 2 WARNINGS: 18 FAILED: 3

1. Error: mlflow can save model function (@test-model.R#24) 

2. Error: mlflow can log model and load it back with a uri (@test-model.R#75) 

3. Failure: mlflow can serve a model function (@test-serve.R#48) 
```

You can see more details at https://travis-ci.org/mlflow/mlflow/jobs/537192519 including full stack trace. I don't see how the ONNX PR affects these tests and have noticed similar failures in some other PRs submitted that also don't change R-related code. Are some tests unstable in R?

To test this PR on travis, I have commented out the R tests and the PR is passing all the checks (See [https://travis-ci.org/mlflow/mlflow/builds/537197979](url).

dbczumar commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,194 @@
+
+class _OnnxModelWrapper:
+    def __init__(self, path):
+        self.onnx_model = _load_model(path)
```
Looks like this is never used. Do we need it?

dbczumar commented on mlflow/onnx.py:
```diff
@@ -0,0 +1,194 @@
+    return onnx_model
+
+
+class _OnnxModelWrapper:
```
@avflor It occurred to me that we can get the input type for each tensor in the ONNX model via the  `onnxrt.InferenceSession` object. Then, we can check for tensors of type `tensor(float)` and cast the corresponding Pandas DataFrame contents accordingly (either a particular column for the multi-input case or the entire set of values for the single-input case).

Here's some code indicating that we can examine the `InferenceSession` object to obtain these types:
```
>>> rt = onnxrt.InferenceSession("my_model.onnx")
>>> rt.get_inputs()[0].type
'tensor(float)'
```

I can make this change if you'd like. Does this make sense from your perspective?

avflor replied:
@dbczumar This is great! Makes the code much cleaner. I quickly made the change and testing it now. I'll let you know if it works and then you can take another pass. Thanks a lot for your help!

avflor replied:
@dbczumar ONNX tests are passing: https://travis-ci.org/mlflow/mlflow/builds/538360806
Now running full tests in travis. Please take a pass and let me know how the code looks like. Also feel free to make any changes :)

dbczumar reviewed (commented):
@avflor I think I may have found a more elegant solution than "try-catch". Let me know what you think!

dbczumar committed:
Tweaks to conform with other flavor conventions

dbczumar committed:
Lint

dbczumar committed:
Add flavor description to Models Docs

dbczumar committed:
Experimental tags

dbczumar committed:
Merge remote-tracking branch 'origin/master' into onnx_tweaks

dbczumar committed:
Mark experimental and docs tweaks

dbczumar committed:
Revert R changes

dbczumar committed:
Merge remote-tracking branch 'origin/master' into onnx_tweaks

dbczumar committed:
Revert spacing changes to Travis yaml

dbczumar committed:
Remove unused arg from get_conda_env

dbczumar committed:
Add test for casting evaluation behavior

dbczumar committed:
Lint

dbczumar committed:
Run tests on xenial to see if onnx import errors remain

dbczumar committed:
Revert travis to trusty and mark onnxruntime-dependent tests as release tests

dbczumar commented:
@avflor @aarondav Our Travis build, which currently runs on Ubuntu 14.04 (Trusty), is currently unable to import the `onnxruntime` library required by the `pyfunc` representation of an MLflow model with the ONNX flavor. Accordingly, I've applied the `release` pytest label to all tests that explicitly reference the `pyfunc` representation. Other tests have been modified to use custom conda environments without the `onnxruntime` library. Provided that there is approval to merge this PR in its current state, I will follow up with a separate PR that removes the `release` label from relevant tests once our Travis build has been updated to support `onnxruntime`.

dbczumar reviewed (approved):
LGTM! Thanks for this contribution @avflor!

avflor commented:
Thank you @dbczumar for your detailed comments and your help! 

Regarding the failed imports on travis -- I wondered what changed in the environment since the imports were successful when I tested a few days ago.
